[DO NOT MERGE] Tracking Devel

csarofeen opened the pull request.

jjsjann123 commented on torch/csrc/jit/passes/constant_pooling.cpp:
```diff
@@ -4,6 +4,7 @@
 #include <torch/csrc/jit/ir/alias_analysis.h>
 #include <torch/csrc/jit/ir/ir.h>
 #include <torch/csrc/jit/ir/node_hashing.h>
+#include <torch/csrc/jit/jit_log.h>
```
We could revert this. :stuck_out_tongue:

jjsjann123 commented on torch/jit/_script.py:
```diff
@@ -1301,6 +1301,10 @@ def forward(self, a) -> MyModule:
             obj = obj.__original_fn
             _rcb = _jit_internal.createResolutionCallbackFromClosure(obj)
 
+        # some functions are explicitly marked as not supported in script mode
```
oops, this should go to upstream!

jjsjann123 reviewed (commented):
minor comments for myself to clean up the merge PR

csarofeen commented on torch/csrc/jit/codegen/cuda/arith.cpp:
```diff
@@ -1274,6 +1316,14 @@ TensorView* reductionOp(
+  // In PyTorch, reduction of a size-0 tensor is effectively creating a tensor
+  // filled with the init value.
+  auto maybe_full =
+      maybeFullInsteadOfReduction(uint_axes, init, tv, keep_dim, dtype);
```
PyTorch, why are you so strange?

jjsjann123 committed:
Merge remote-tracking branch 'upstream/viable/strict' into HEAD

naoyam committed:
Refactor lower_alias_memory.cpp (#2170)

* Refactoring of lower_alias_memory

csarofeen committed:
Merge pull request #2172 from csarofeen/upstream_bump_1109

Upstream bump 1109

zasdfgbnm committed:
Misc mutator cleanups/fixes (#2180)

jjsjann123 committed:
fixing my python flake8 issue from merging warning (#2182)

removing label workflow

naoyam committed:
Fix invalid aliasing (#2178)

* Look at all the loops rather than just the consumer IDs as there can be
loops not mapped to the consumer.

zasdfgbnm committed:
Add support for `select` op (#2179)

zasdfgbnm committed:
More trivial reduction cleanup (#2181)

azazhu committed:
Print debug info about expr transform on GpuLower (#2185)

* add print debug for nvfuser

* refine dump exprs code: 1) rename option name 2) move reduplicated logic to func dumpExprsIfEnabled

Co-authored-by: Feiwen Zhu <mzhu@nvidia.com>

zasdfgbnm committed:
Reduce the work to add a new expr: remove ExprType (#2186)

naoyam committed:
Add computeWith to interleave gmem accesses and computations (#2156)

Add computeWith to interleave gmem accesses and computations

liqiangxl committed:
bug-3607940: fix wave quantization and round bdimx in outer reduction based on problem size (#2191)

xwang233 committed:
Add support for normal distribution RNG (#2171)

naoyam committed:
Add Float IR node class (#2197)

* Add Float IR node class

Represents the 32-bit floating-point scalar value. Not supported in
PyTorch, so can't be used as inputs to fusions

zasdfgbnm committed:
Reduce the work to add a new expr: unify the structure of exprs (#2190)

naoyam committed:
Support int32 Int (#2198)

naoyam committed:
Don't set DataType of IterDomain as Int (#2200)

naoyam committed:
delete unused function (#2201)

zasdfgbnm committed:
Reduce the work to add a new expr: mutator cleanup (#2199)

naoyam committed:
Refactor scalar IR nodes (#2203)

* Refactor scalar IR nodes (Int, Double and Bool)

Everything uses template class Scalar

zasdfgbnm committed:
Reduce the work to add a new expr: graphviz cleanup (#2205)

naoyam committed:
Suppress nvcc warnings (#2206)

liqiangxl committed:
use reference, some compilers may complain or treat it as error (#2211)

zasdfgbnm committed:
Reduce the work to add a new expr: rewrite SubstituteInExpr with OptOutMutator (#2202)

naoyam committed:
Vectorized welford (#2204)

* "Vectorize" sequential welford computations

Lift the predicated count division outside of the innermost loop if that
loop is exactly mapped with vectorized IDs and not a reduction domain.
Targeted to address outer-reduction grid welford tuning

naoyam committed:
Add missing clone (#2213)

naoyam committed:
WAR to make nvcc compile generated code (#2214)

The complex type defined by `defineComplexType` doesn't seem to work
with nvcc. When compiled with nvcc, instead of using the definitions,
include <complex>. Also disable some other parts that don't work with
<complex>.

This would of course break the complex type support, but as long as it's
not used, it shouldn't matter. Compilation with nvcc should be just
ad-hoc experiments of code modifications, so the limitation should be
fine.

csarofeen committed:
Rework vectorization analysis for View (#2124)

Disable conservative vectorization validation.
Cache vectorize mapper use in heuristics.
Change vectorization analysis to be fractional and divisibility based.
Extract out extent propagation.
Vectorization pass implements its own spanning tree.
If any vectorized tensor doesn't have a mapped dimension return size 1 vectorize.

naoyam committed:
Fix expr sorting (#2217)

Originally fixed in the upstream PR #89551

csarofeen commented on torch/csrc/jit/codegen/cuda/lower_expr_sort.cpp:
```diff
@@ -418,60 +418,61 @@ std::vector<ExprGroup*> ExprGroup::getMergeCandidates(
   }
 
-  std::vector<bool> can_merge(true, neighbors.size());
+  std::vector<bool> can_merge(neighbors.size(), true);
```
🤦‍♂️

azazhu committed:
Add support for index_select OP (backend implementation) (#2183)

* add index_select(lookup_tv, dim, index_tv) fusion for nvfuser: 1) support arbitrarily rank; 2) support arbitrarily dim value; 3) index_tv can become consumer; 4) lookup_tv cannot become consumer. TODO: move index calculation from codegen.cpp to lower_index.cpp

* Add shallowcpy implementation for IndexSelectOp

* 1) remove print debug feature and frontend features to another tow PRs
2) mv test_index_select.cpp to torch/csrc/jit/codegen/cuda/test/
3) refine code

* mv test_index_select.cpp

* refine code

* mv lower_index from codegen.cpp to lower_index.cpp

* refine

* 1. refine UT by adding seed 2. add getSelectedDomainIfTvIsIndexSelectOutput 3. add func in validateIr

* refine format

* rename func maybe_broadcast_index_tv

* fix IrPrinter::handle(IndexSelectOp)

Co-authored-by: Feiwen Zhu <mzhu@nvidia.com>
Co-authored-by: Chang Liu <chang.liu@utexas.edu>

csarofeen committed:
Move tests around. (#2208)

naoyam committed:
Remove dummy outputs before inlining (#2219)

Fixes #2212

azazhu committed:
Add support for index_select OP (frontend implementation) (#2184)

* add index_select frontend

* 1. use EnableOption::GraphOp to control index_select OP parse 2. comment out call index_select func as it is in backend PR

* 1. add profiled-constant 2. add an entry here for type inference 3. add e2e py

* 1. remove code for legacy fuser 2. add assert on index_select func

* 1. refine index_select e2e test 2. add EnableOption::GraphOp's token

Co-authored-by: Feiwen Zhu <mzhu@nvidia.com>

csarofeen committed:
Minor cleanup, mostly of Transform Replay (#2220)

naoyam committed:
Buffer overrun detected with asan (#2218)

Also changed [i] to .at(i)

csarofeen committed:
Don't exact map across data swizzles. (#2222)

naoyam committed:
Add a unit test using an external source file (#2226)

I often create this unit test in my local branch to play around using
manually modified generated code. Giving it a permanent place as it may
be useful for anyone working on code generation.

naoyam committed:
Make sure predicate elimination info propagated through unarySetInserter (#2136)



Co-authored-by: Ryan Spring <rdspring1@gmail.com>

naoyam committed:
Cleans up the option handling code for PYTORCH_NVFUSER_DUMP/DISABLE/ENABLE (#2225)

* Cleans up the option handling code for PYTORCH_NVFUSER_DUMP/DISABLE/ENABLE

Fixes missing initialization of DisableOption::GraphOp.

csarofeen committed:
Fix comments due to PR #2208 (#2223)

* Don't exact map across data swizzles.

* Fix comments.

csarofeen committed:
Cleanup some nonsymmetric BestEffortReplay usage. (#2221)

chang-l committed:
Add performance benchmark for indexselect OP  (#2224)

* Add indexselect benchmarks

* Update indexselect benchmark

* Increase size to better measure bare indexselectOp

* Rename tests IndexSelectFused -> IndexSelect

kevinstephano committed:
Add support for a symbolic output_shape for broadcast_in_dim (#2195)

* Add support for a symbolic output_shape for broadcast_in_dim in the Python Frontend

* Fixed compilation of BroadcastInDimOpRecord with template specialization for defining expand sizes.  Added SymbolicSizesRecord.  Added python binding for symbolic_sizes().

* Fix up code with some suggestions from Ivan.  Changed symbolic_sizes() to tensor_sizes().

* Add some tests and fix up python based Tensor and Scalar printing.

* Added testing for tensor_sizes(). Made some minor changes to faciliate string captured definition testing.

* Add comments and fix lint issues.

* Add an output broadcast test with tensor_sizes().

* Added a test for tensor_size usage when each operand of a binary op has a broadcast.

* Fix tensor_sizes() to reflect an expand in extent.  Add appropriate test.

naoyam committed:
Allow IrBuilder to have Int and Int32 inputs in binary expressions (#2228)

naoyam committed:
Fix #2230 (#2231)

zasdfgbnm committed:
A new index hoisting and CSE approach based on subexpression equivalence and dependency (#2235)

zasdfgbnm committed:
Change TensorIndex to store a single Val* (#2234)

jjsjann123 committed:
disable index_select fusion for 0-dim input (#2247)

patch for fusion logic on index_select to disable fusion for 0-dim input tensor

liqiangxl committed:
empty gpu memory cache between different benchmark cases (#2242) (#2243)

zasdfgbnm committed:
Change indexing of tensor factories (#2248)

zasdfgbnm committed:
Hoist entire predicates (#2249)

zasdfgbnm committed:
Use non-broadcasting positions in `multiReductionInliner` (#2250)

naoyam committed:
Grouped grid welford optimized for outer reduction (#2229)

* Add welfordGroupOuter

Grouped grid welford specialized to outer reductions where TIDx and BIDx
are used to parallelize a non-reduction inner domain, and TIDy and BIDy
are used to parallelize a reduction outer domain.

This is a very specific version, but for intended cases it should
perform much better than the default welfordGroup. Two major factors are
the faster block welford using warp shuffle instructions and reduced
register usage by avoiding complex template parameters. The latter
optimization can be applied to other reduction patterns, but not
explored yet.

* Drop the outer-opt block welford

This was the first attempt to optimize grouped welford. Was fairly
effective but required the shmem buffer size that is larger by a factor
of the group count. For large group counts (>= 4), the new
implementation is more efficient.

zasdfgbnm committed:
Hoist entire index (#2255)

zasdfgbnm committed:
Scalar hoist self-reusing (#2254)

zasdfgbnm committed:
Expr simplifier step 1: mathematical theory and associative and commutative op reordering (#2215)

zasdfgbnm committed:
Make expr sort put scalar exprs in the front to enable more hoisting opportunity (#2252)

zasdfgbnm committed:
Move IR printing definition to the class, step 1: Val (#2256)

naoyam committed:
Minor refactoring of scheduler (#2253)

* Minor cleanup. Originally done in #1772

* Revert the change on the number of used SMs.

Unclear what (warp_size / 8) * device_multiprocessor_count means, but
this PR shouldn't change the current heuristics

naoyam committed:
Update and fix comments (#2261)

zasdfgbnm committed:
Simplify loop extent in codegen (#2257)

liqiangxl committed:
bug-3607824: add maxregcount in innerPersistentHeuristic to increase … (#2233)

* bug-3607824: add maxregcount in innerPersistentHeuristic to increase occupancy

* refactor maxrregcount

* set benchmark

* optimize buffer size and register

* fix comments

* fix executor, add comments to normalization scheduler

* recompile if maxrregcount changed

* adjust pb and bdimx

* small block large buffer

* fix comments

* set maxthreads to 1024, allow pb=10

* reduce pb by half

* sset max threads to 1024 only if there is no outer reduction

* add comment

* skip set bdimx for cases have outer reduction

zasdfgbnm committed:
Fix arg eval order (#2265)

zasdfgbnm committed:
Move IR printing definition to the class, step 2: Exprs (#2263)

zasdfgbnm committed:
ExpressionEvaluator move op impl to Expr::evaluate (#2262)

zasdfgbnm committed:
Add debug printing for expr simplifier (#2267)

zasdfgbnm committed:
Allow debug dump, enable, and disable options to have arguments (#2268)

zasdfgbnm committed:
expr simplifier: eliminateTrivialComputation (#2266)

zasdfgbnm committed:
ir_utils::varName (#2271)

zasdfgbnm committed:
Expr simplifier: prove::isNonZero (#2270)

zasdfgbnm committed:
Allow users to specify which pass to trigger PYTORCH_NVFUSER_DUMP="expr_simplify" printing (#2269)

chang-l committed:
Add auto-grad support for index_select op (#2259)

* Add index_select in diff group

* Update to make it recognizable in diff group

* Add python unit test

* Address comments

zasdfgbnm committed:
Fix CI (#2274)

zasdfgbnm committed:
FlattenedAssocCommOp changes (#2272)

ftxj committed:
Add support for torch.gather OP (backend implementation) (#2232)

* add support for torch.gather (backend)

* add TorchScatter IterType

* Support index_tv in torch.gather become consumer

* Fix TorchGatherOp.dim() bug

* Add comments for gather op

* Add new lines at end-of-file

* Add accessors

* WIP: make indexing work even with different-sized IterDomains

* Fix root map with torch_gather

* Remove unused function

* Fix gather iteration domain merge bug

* Fix torch.gather toString method

Co-authored-by: Naoya Maruyama <nmaruyama@nvidia.com>

zasdfgbnm committed:
Fix indexing when swizzle exist (#2283)

zasdfgbnm committed:
Fix lower_verbose printing (#2281)

jjsjann123 committed:
Merge commit 'd19988093d1b7942d15dce6ef61b62dccfc3b8a3' into HEAD

csarofeen committed:
Merge pull request #2289 from csarofeen/upstream_merge_1221

Upstream merge 1221

jjsjann123 committed:
Nvfuser build refactor cherry pick (#2292)

* [NVFUSER] refactor nvfuser build

Cherry-picking upstream PR: https://github.com/pytorch/pytorch/pull/89621

This PR is the first step towards refactors the build for nvfuser in order to have the coegen being a standalone library.

Contents inside this PR:

    nvfuser code base has been moved to ./nvfuser, from ./torch/csrc/jit/codegen/cuda/, except for registration code for integration (interface.h/interface.cpp)
    splits the build system so nvfuser is generating its own .so files. Currently there are:
        libnvfuser_codegen.so, which contains the integration, codegen and runtime system of nvfuser
        nvfuser.so, which is nvfuser's python API via pybind. Python frontend is now exposed via nvfuser._C.XXX instead of torch._C._nvfuser
    nvfuser cpp tests is currently being compiled into nvfuser_tests
    cmake is refactored so that:
        nvfuser now has its own CMakeLists.txt, which is under torch/csrc/jit/codegen/cuda/.
        nvfuser backend code is not compiled inside libtorch_cuda_xxx any more
        nvfuser is added as a subdirectory under ./CMakeLists.txt at the very end after torch is built.
        since nvfuser has dependency on torch, the registration of nvfuser at runtime is done via dlopen (at::DynamicLibrary). This avoids circular dependency in cmake, which will be a nightmare to handle. For details, look at torch/csrc/jit/codegen/cuda/interface.cpp::LoadingNvfuserLibrary

Future work that's scoped in following PR:

    Currently since nvfuser codegen has dependency on torch, we need to refactor that out so we can move nvfuser into a submodule and not rely on dlopen to load the library. @malfet
    Since we moved nvfuser into a cmake build, we effectively disabled bazel build for nvfuser. This could impact internal workload at Meta, so we need to put support back. cc'ing @vors

naoyam committed:
Add outer normalization tests with perf measurement switch (#2236)

* Add outer normalization tests with perf measurement switch

Set PYTORCH_NVFUSER_OUTER_REDUCTION_BENCHMARK=1 to enable timing
measurements

Scheduling parameters are tuned on Titan RTX and are not optimal on
other devices. The purpose of this PR is to add tests of the scheduling
pattern implemented in `scheduleNormalization`, which is going to be
used as the reference for developing a new outer normalization scheduler

naoyam committed:
Add a missing override (#2296)

naoyam committed:
Make PairwiseRootDomainMap require the same extent by default (#2298)

naoyam committed:
Allow -maxrregcount value to be overwritten by (#2299)

PYTORCH_NVFUSER_MAX_REG_COUNT

Also cleaned up the related code

jjsjann123 committed:
Code cleaning (#2293)

code cleaning
fixing sign-comparator

jjsjann123 committed:
Fixes nvfuser installation and examples (#2303)

csarofeen committed:
nvFuser benchamrk moving to third_party/nvfuser (#2294)

moving nvfuser_bench under third_party/nvfuser
updating directory for benchmark

Co-authored-by: jjsjann123 <alex.jann2012@gmail.com>

kevinstephano committed:
Kernel Database (redo of PR #2193) (#2297)

* Adding kernel_db code to third_party nvfuser code.

* Made a fix to properly destroy nvrtc Program. Also, made a fix to properly set disable of kernel_db.

kevinstephano committed:
Changing FusionCache names in the Python Frontend to reflect that the underlying data structure is a trie. (#2304)

naoyam committed:
Build fix (#2305)

naoyam committed:
clang-format (#2306)

jjsjann123 committed:
Lintrunner compiler warning patch (#2312)

fixing compiler warning
fixing lintrunner entries

naoyam committed:
Revert CMakeLists.txt as build fails with clang due to unused variables (#2315)

kevinstephano committed:
Modify `FusionDefinition.add_output` API to an include an `alias_input` option. (#2314)

* Add alias_output_to_input API to Python Frontend.

* Make aliasOutputToInput as an option of add_output in python frontend.

* Add test for fd.add_output() alias_input option.

* Fix lint issues in test.

* fixing nvfuser alias output entry cache for python frontend (#2319)

* Removed ability to alias an input for scalar.  Fixed up tests for aliasOutputToInput so they test for aliasing of not the aliased inputs surrounded by outputs.

Co-authored-by: jjsjann123 <jiej@nvidia.com>

liqiangxl committed:
add enable option to warn register spill (#2313) (#2317)

liqiangxl committed:
replace welford in layer_norm forward with two-pass (#2300) (#2302)

jjsjann123 committed:
fixing segfault on CI. The order of argument elimination was wrong (#2323)

fixing segfault on CI. The order of argument elimination was wrong
clangformat

jjsjann123 committed:
Code cleaning (#2320)

Revert "Revert CMakeLists.txt as build fails with clang due to unused variables (#2315)", so more compiler warning are enabled.
fixing compiler warnings. Verified on both gcc & clang-14

jjsjann123 committed:
removing wrong check in kerneldb test (#2324)

Fixes CI failure.
non-existing db path shouldn't fail

liqiangxl committed:
check welford input before trial translate to two-pass (#2325)

jjsjann123 committed:
patching failing segfault in test_jit (#2328)

Fixes segfault from running ./build/bin/test_jit

Not totally sure what the root cause is though. The issue is tracked in: https://github.com/csarofeen/pytorch/issues/2329

jjsjann123 committed:
adding a -Wall in case compile options passed from dependency changed (#2330)

zasdfgbnm committed:
Add FusionExternalSrcMatmul_CUDA to test_gpu_external_src.cpp (#2287)

zasdfgbnm committed:
Expr simplifier: divisibility analysis (#2258)

naoyam committed:
Do not warn unused functions (#2333)

In clang, options placed later overwrites options placed before, so
having just -Wall here means all -Wno-*** will be effectively cancelled,
which is different from the behavior in gcc. Since we're okay with
unsued functions (at least that's what torch_compile_options did), and
there is unused functions, this flag is necessary in clang.

zasdfgbnm committed:
Allow disabling nvfuser_zero (#2277)

zasdfgbnm committed:
Add comment about swizzle at getGlobalProducerStridedIndices (#2334)

zasdfgbnm committed:
Expr simplifier: implement prove::isPositive, prove::isNonNegative, prove::isNonZero (#2273)

zasdfgbnm committed:
Enable view on TorchScript (#1966)

naoyam committed:
Reduce tensor sizes to avoid OOM on 3080 (#2335)

zasdfgbnm committed:
Expr simplifier: simplification passes for matmul (#2275)

zasdfgbnm committed:
Indexing shared memory with `unsigned` and hoist cvta (#2286)

naoyam committed:
Revert PR #2328 as nothing seems to fail anymore (#2339)

Closes https://github.com/csarofeen/pytorch/issues/2329.

Will re-open if the error happens again.

ftxj committed:
Add support for torch.gather (frontend) (#2280)

This PR is frontend implementation of torch.gather(lookup_tv, dim, index_tv, *, sparse_grad = False) fusion

Co-authored-by: Naoya Maruyama <nmaruyama@nvidia.com>
Co-authored-by: jjsjann123 <alex.jann2012@gmail.com>
Co-authored-by: jjsjann123 <jiej@nvidia.com>

liqiangxl committed:
simplify TensorDomain::noReductions and TensorDomain::noBroadcasts using std::copy_if (#2308) (#2311)

zasdfgbnm committed:
Expr simplifier: predicate register usage optimization (#2276)

zasdfgbnm committed:
Fix position of toSmem for scalar hoisting (#2340)

zasdfgbnm committed:
Remove obselete `#if defined(USE_CUDA)` (#2341)

zasdfgbnm committed:
Disable nvfuser zero for matmul (#2342)

jjsjann123 committed:
patching reshape with scalar tensor input (#2343)

Fixes reshape CI failure.

Disable TorchScript fusion of reshape op with scalar tensor input, which triggers an assertion failure on original input shapes.

jjsjann123 committed:
Revert "Revert PR #2328 as nothing seems to fail anymore (#2339)" (#2346)

This reverts commit e08f2c181a022a1ae9523b7e3ee5b29794a3e223.

The segfault went back.

kevinstephano committed:
Remove `FusionInterface` class (c++) and `Fusion` class (python) from Python Frontend API (#2337)

* Adding a container for FusionExecutorCache inside of FusionCache in the Python Frontend.

* Make FusionCache work with new FusionSchedulers struct.

* Remove some un-needed comments.

* Add SchedOperators to hold python scheduling operators to py bindings.

* Add function based definition and schedule.

* Remove FusionInterface class and consilidate functionality into FusionDefinition.

* Fixed python_bindings.cpp include of fusion_interface.h file that needed to be removed.

* Remove more of the usage of the nvFuser Python Fusion object.

* Add python directory and fix lint issues.

* Fix issues from review. Fixed imports that use ._C.  Moved README to python directory.

* Fixed Lint issues.

* Removing some FusionDefinition methods that I did not mean to add.  Fixed import path in python/__init__.py as python tests were failing.

kevinstephano committed:
Fix gather test to include removal of Fusion object usage. (#2349)

liqiangxl committed:
Disable reuse of x_mean_sub in translateSingleWelford (#2322)

jacobhinkle committed:
Fix ExpandOp::toString printing leading comma (#2338)

jjsjann123 committed:
Index select patch (#2353)

Things in this PR:

    fixes python frontend exposure of index_select.
    add python frontend test for index_select.
    cleans up boilerplate of scheduler for select/index_select/gather.
    added fusion input check for select/index_select/gather ops in no-op scheduler. (Fixes part of 

SegmentCandidateFinder (fusion_segment.cpp) asserts on findSegments() #2352 on the inconsistent error message for slow tests. Note that the real fix also would requires the segment refactor in [WIP] Insert floating-point type casting at every segmentation step #2344 )

zasdfgbnm committed:
Validate the corretness of results when running matmul benchmarks (#2356)

naoyam committed:
Reinitialize segments after welford translation (#2350)

* Reinitialize segments after welford translation

Closes #2348

naoyam committed:
Insert floating-point type casting at every segmentation step (#2344)

* Insert floating-point type casting at every segmentation step

It is currently only done after completing the segmentation. This could
lead to inconsistent scheduling decisions as type casting could lead to
different scheduling decisions due to, for example, smaller persistent
buffer sizes.

naoyam committed:
Do not segment UnaryOps from fusion inputs (#2354)

* Do not segment UnaryOps from fusion inputs

zasdfgbnm committed:
Add PYTORCH_NVFUSER_DISABLE="expr_simplify(pass,names)" (#2363)

zasdfgbnm committed:
Add debug print for expr sorting (#2365)

zasdfgbnm committed:
Improve printting of LoadStoreOp and MMAOp (#2364)

naoyam committed:
Fix resetting expr list of SegmentedGroup (#2368)

naoyam committed:
Do not insert fp16 cast if used by select-like ops (#2357)

xwang233 committed:
Add randn, randn_like (#2355)

zasdfgbnm committed:
this test is NOT failing! (#2371)

naoyam committed:
fix (#2385)

jjsjann123 committed:
python tests moving under nvfuser directory (#2373)

moving python test scripts from `./test` to `./third_party/nvfuser/python_tests/` with renaming
updating cmake files to create a symbolic link to `build_dir/build`, so we can run nvfuser_tests in `manual_ci.sh` 
installing nvfuser python test scripts under `test/_nvfuser`
place_holder for old nvfuser related tests added back to pytorch core `./test`

Note:
1. From now on, everyone should run `./third_party/nvfuser/manual_ci.sh` before merging!
2. skip failing dynamo test which has been patched in upstream, we need to revert this skip once we are build against upstream/master.

naoyam committed:
Fix projected size calculation (#2382)


* Fix #2381

zasdfgbnm committed:
Reduce tensor size in FusionExprSortMatmulLikeSchedule_CUDA (#2395)

jjsjann123 committed:
Added python API `nvfuser.version()` (#2393)

Fixes #2379

Note: developers are supposed to bump the version number (./third_party/nvfuser/version.txt) when user facing APIs are updated.

nvfuser.version() returns a three digit version plus 7-digit sha from git commit. where it

returns nvfuser version in format of a string 'm.n.p+git[7d-sha]'.
We strip the git[7d-sha] and convert the string to
`packaging.version.Version` for comparison. e.g. you can use it as:
    import nvfuser
    print(nvfuser.version())              # 0.0.1+git21df524
    nvfuser.version() == '0.0.1`          # True
    nvfuser.version() > '0.0.0`           # True
    from packaging.version import Version
    nvfuser.version() < Version('1.0.0')  # True
Currently the three digit version is hard coded inside third_party/nvfuser/version.txt. While the git sha is retrieved inside cmake, and the version string is then constructed and written to thrid_party/nvfuser/python/version.py

TODO:
We'll want to have something automatically in there to enforce API compatibility in the future.

kevinstephano committed:
Add `from_pytorch` function to `FusionDefinition` API and remove number of dims based `define_tensor` (#2375)

* Add empty_tensor to the FusionDefinition API.  Remove the num dims based define_tensor function.  Add Python based FusionDefinition class.

* Add an API to `FusionDefinition` to allow for child classes to define fusions through a function (#2376)

* Add function based FusionDefinition.

* Add FusionDefinition func definition based tests.

* Fix lint issues.

* Removed an empty line from __init__.py file and fixed a badly formed python function test.

* Forgot to remove nvfuser._C from test_prims.py

kevinstephano committed:
Rename view to reshape. (#2397)

jacobhinkle committed:
Support 0-dimensional shapes as inputs/outputs to view() (#2398)

* Add tests for 0-dim input&output shapes in view

* Add special-case for 0-dim tensors in analyzeView

This just short-circuits analyzeView for the special case of
0-dimensional input shape. 0-dimensional output shape seems to work fine
without this change.

* Test mixed bcast/concrete dims in view input

* Rename view->reshape

cf. PR 2397

---------

Co-authored-by: Jacob Hinkle <jhinkle@nvidia.com>

zasdfgbnm committed:
FusionExprSortMatmulLikeSchedule_CUDA use double (#2401)

zasdfgbnm committed:
Add `FusionExecutor::disassembledBinary` (#2402)

jjsjann123 committed:
updating scripts for ci test (#2389)

jjsjann123 committed:
Copy variant cleanup (#2404)

Removing the use of prim::xxx_copy variants that has already been added to aten.
This also resolves failures on upstream/master pytorch core due to the lack of overloaded variants that's not in our registered prim ops.

jjsjann123 committed:
Nvfuser version patch (#2407)

quick patching for nvfuser version comparison to support `packaging.version.Version`

jacobhinkle committed:
Allow cacheAfter for inputs that are outputs (#2374)

* Allow cacheAfter for inputs that are outputs if they are used

* Use makeContigTensor in test FusionIssue2372_CUDA

* Replace addOutput(input) with call to aliasOutputToInput

This commit alters the order of aliasOutputToInput slightly to give
some state that can be checked inside addOutput in order to avoid
infinite recursion.

* Only check for uses (not isFusionOutput) in cacheAfter

---------

Co-authored-by: Jacob Hinkle <jhinkle@nvidia.com>

naoyam committed:
Misc changes/cleanups (#2408)

* Misc changes/cleanups

liqiangxl committed:
allow load external cuda code from python (#2411)

jacobhinkle committed:
Remove automatic aliasOutputToInput in addOutput (#2414)

This caused a lot of breakage since aliased outputs are not returned as
outputs. Rather than just fixing the tests, it may be better to handle
the special aliasing cases separately.

Co-authored-by: Jacob Hinkle <jhinkle@nvidia.com>

zasdfgbnm committed:
Some more about SASS printing/checking (#2415)

naoyam committed:
Refactor vectorization in the schedulers (#2412)

* Refactor vectorization in the schedulers

zasdfgbnm committed:
Some C++17 cleanup (#2420)

zasdfgbnm committed:
DisableOption::IndexHoist should not disable index simplification (#2426)

ftxj committed:
[Scatter 2] Add ScatterOp IR data structure (#2391)

naoyam committed:
Print float constant values as "...f" (#2427)

rdspring1 committed:
Refactor squeeze prim to accept a sequence of dimensions. (#2428)

zasdfgbnm committed:
Add new op: iota (#2433)

zasdfgbnm committed:
cp.async.cg (#2429)

jacobhinkle committed:
Add dtype argument to define_constant (#2423)

* Add FloatConstantWhere test

This test ensures that given two Float constants, where infers the
output dtype properly.

* Add missing ComplexFloat cases in dispatch.cpp

* Add frontend tests using constants with dtype argument

The complex constant tests currently fail with a TypeError.

* Check dtypes in FusionFloatConstantWhere_CUDA

* Add dtype and fix complex printing in ConstantRecord

This fixes the issues I was having with defining complex constants. The
problem was that the Python frontend actually generates Python code
which gets evaluated, and that Python code was simply printing the
values assuming they were real (if not bool). In the case of
std::complex, C++ prints those as pairs i.e. a+b*j is printed as (a, b),
which in Python is a tuple, not a complex value. Instead, we now print
them as a+bj which is the compact form for a Python complex literal.

* Enable complex constant tests in test_python_frontend.py

---------

Co-authored-by: Jacob Hinkle <jhinkle@nvidia.com>

zasdfgbnm committed:
Simplify expr with the fundamental division-with-remainder property (#2434)

zasdfgbnm committed:
Assert LDSM register usage (#2431)

zasdfgbnm committed:
Predicate simplification: `b && b`, `b || b` (#2436)

liqiangxl committed:
avoid promote float16 and bfloat16 to float32 in op-where (#2409)

ftxj committed:
Add index_select shape expression (#2418)

* Fix index_select shape expression

* rename infer_index_select_size func

* Add index_select test

* Add assert

* Add FileCheck for index_select_shape_expr

zasdfgbnm committed:
Fix cpAsyncCg (#2439)

naoyam committed:
Just to make nvcc happy with generated code (#2444)

naoyam committed:
Fix input type lowering in segmentation (#2416)

* Fix input type lowering in segmentation

jjsjann123 committed:
switching `packaging.version.Version` to `torch.torch_version.Version` (#2422)

1. Avoids nvfuser importing error on platforms without python packaging. copied LazyImport from torch_version.py
2. Patch on confusing error message when ./nvfuser/__init__.py fails on consecutive imports.

zasdfgbnm committed:
Pull some tests from Shiming's PR (#2446)

jacobhinkle committed:
Add complex rsqrt function to runtime/helpers.cu (#2448)

* Add complex rsqrt function to runtime/helpers.cu

* Fix rescaling in rsqrt helper functions

* Cleanup in rsqrt

* Fix namespace issue in complex rsqrt

* Add real rsqrt helpers.

This fixes compile errors in test_torchscript.py, specifically test_unary_ops
and test_disable_const_chunk_propagation_for_normalization that were due
to missing overloads for the new rsqrt in the codegen scope.

---------

Co-authored-by: Jacob Hinkle <jhinkle@nvidia.com>

jjsjann123 committed:
Scalar type patch (#2452)

Construct the correct Val storage type for complex scalar.

zasdfgbnm committed:
Add python binding for iota (#2447)

zasdfgbnm committed:
Convert aligned block sync (#2445)

zasdfgbnm committed:
Temporarily disable cp.async.cg for now (#2457)

zasdfgbnm committed:
Fix SASS test, manually set PATH (#2462)

naoyam committed:
Clear thread predicates when RAW sync is used (#2460)


* Clear thread predicates when RAW sync is used

naoyam committed:
Enable grid outer persistent scheduling (#2435)

* Enable grid outer persistent scheduling

naoyam committed:
Generate where with C++ `? ... : ...` rather than using device functions (#2472)

naoyam committed:
hasToString util template function to print using toString (#2471)

zasdfgbnm committed:
Fix cp.async barrier (#2470)

zasdfgbnm committed:
Improve error message for SASS dumping (#2474)

naoyam committed:
Move arith to ops directory (#2475)

* Move arith.h/cpp to ops

* Moved util functions in ops/arith.cpp to ops/utils so that they can be
also used from other ops functions

naoyam committed:
Build fix (#2476)

* Build fix

zasdfgbnm committed:
Move our codebase to namespace nvfuser (#2477)

zasdfgbnm committed:
Skip FusionCpAsyncCommitWait_CUDA on pre Ampere (#2481)

zasdfgbnm committed:
Swap order of stdout and stderror read from nvdisasm (#2483)

mmigdal-nv committed:
Moving predicate in cp.async (#2484)

naoyam committed:
Thread predication of reduction init expr (#2489)


* Fix #2487

zasdfgbnm committed:
Don't initialize if using cpAsync (#2485)

zasdfgbnm committed:
Fix buffer reusing (#2490)

liqiangxl committed:
set heuristics to persistent after welford translatte (#2491)

zasdfgbnm committed:
Cleanup debugging print (#2497)

zasdfgbnm committed:
Return Bool* for logical expr in IrBuilder (#2494)

zasdfgbnm committed:
Make IrCloner more convenient to use (#2495)

kevinstephano committed:
Add Minimal User Scheduling to Python API (#2438)

* Add schedule function to FusionDefinition python api.

* Add user schedule functions.

* Add queryUserSchedule() to FusionCache.

* Add User Schedule query methods to FusionCache and use them in FusionDefinition::execute()

* Defined FusionDefinition setupSchedule and finalizeSchedule methods.

* Add createUserSchedule method to FusionCache.

* Allow user to define an empty schedule for a definition.

* Fix comments in FusionDefinition.

* Add split and merge operations as well as re-build Fusion IR with basic definition.

* Cleanup FusionDefinition comments and code related to user scheduling.

* Add checking for definition prior to use SchedOperators ops.

* Add Operator error checking to prevent them from being used during user schedules.

* Fix lint issues.

* Remove func_based_def data member from FusionDefinition exposure in python.

* Fix FusionDefinition data member name usage.

* Fixed const-ness of returning a UserSchedule on query.  Also added mutex guards to cache queries and node creation.

* Fixed lint issues.

* Fix split python API and add (non-working) test.

* Fix setupSchedule() that was broken due a bad merge in buildFusionIr().  Added a test to schedule split and merge as a sanity check.

* Fixed an error check!

* Moving FusionDefinition.execute() additional args to kwargs.  Removed def_complete data member from FusionDefinition C++ class in favor of using fusion_id_.

zasdfgbnm committed:
Lower unroll cleanup, make it support IfThenElse (#2496)

zasdfgbnm committed:
Debugging helper: assert when accessing global tensor out of bound (#2493)

kevinstephano committed:
Fix errors resulting from fusions with scalar only inputs (#2503)

* Fix getCommonDeviceCUDA() when there is only a scalar input such that it returns a non-failure indicating result.

* Add a test for scalar only inputs.

zasdfgbnm committed:
Fix canOmitStopPredicate (#2504)

jjsjann123 committed:
patching dynamo test (#2498)

Quick fix on dynamo import in python tests. Patches dynamo test failure with upstream's updated API

zasdfgbnm committed:
Refactor DataType to better support complicated dtypes like array and pointer (#2417)

liqiangxl committed:
disable welford translate if the original scheduler before translate is not persistent (#2509)

jacobhinkle committed:
Fix mismatched signature in addcmul macro (#2512)

Fixes #2506

Co-authored-by: Jacob Hinkle <jhinkle@nvidia.com>

ftxj committed:
[Scatter 3] Add IndexLowering for ScatterOp (#2437)

* Add consumer index compute using override_index

* Add override param for lowerDst

* Add override param for lowerDstIndex

Make the contig indexing ignore GatherScatter IDs



---------

Co-authored-by: Naoya Maruyama <nmaruyama@nvidia.com>

jjsjann123 committed:
index_select/gather cache patch (#2501)

patching index_select/gather on cache check to examine reduction dimension. (original issue exposed lightning)
Added python tests to verify the fix.
lintrunner to check FLAKE8 on nvfuser python code

ftxj committed:
Move zero shape tests (#2514)

rdspring1 committed:
Refactor `FusionDefinition` to support deserialization. (#2507)

* Create separate base class FusionState from FusionDefinition
* Move struct State to fusion_state.h

naoyam committed:
Bug fix (#2518)

naoyam committed:
cleanup (#2519)

naoyam committed:
cleanup (#2520)

mmigdal-nv committed:
Adding current stream to event record which (#2521)

jacobhinkle committed:
Add frontend arithmetic ops and shape property (#2479)

* Add fd member to Tensor and Scalar

* Add .shape property to Tensor

This also includes a fix for the printing of the TensorSizesRecord, so
that it always prints as a tuple assignment. Without that change, the
test was failing when the printed python definition was evaluated.

* Add unary ops __neg__ and __abs__ to Tensor/Scalar

Since we don't have an inv() op, we can't add __inv__ (i.e. ~x). So I
think these are the only unary ops we have. I will add binary ops next.

* Add special Python binary operators to Tensor and Scalar.

See the note about `__div__` not being implemented. We could implement
it if we need to, but since we don't even have the op currently, I'm
ignoring it for now.

* Rename to _get_fusion_definition

Also removed a junk definition of __add__

* Add Tensor.ndim property to frontend

This allows us to find the dimension of a tensor without creating
new Scalars, as we would if we used `x.shape` or
`fd.ops.tensor_sizes(x)`.
(Also clang-formatted code).

* WIP: Add methods for most ops

I believe this commit is mostly working, but the test segfaults due to a
potential bug in addcmul. I will revisit this after that bug is fixed,
and clean up the tests then.

* Enable addcmul in op methods test

* De-blackify test_python_frontend.py

I had run black on this file to format it, but it makes for a noisy PR.

* Rename {Fusion,Scalar}.fd to .fusion_definition

---------

Co-authored-by: Jacob Hinkle <jhinkle@nvidia.com>

jacobhinkle committed:
Add prod op (#2511)

* Add prod op

* Accept None/scalar dims argument in reduction ops.

These are actually the only signatures accepted by the pytorch
prod/min/max functions.

* Add optional dtype argument to oneVal and zeroVal

---------

Co-authored-by: Jacob Hinkle <jhinkle@nvidia.com>

naoyam committed:
disable a failing test (#2528)

jjsjann123 committed:
Random op python api (#2430)

exposes normal & uniform on python API. Fixes issue #2386

liqiangxl committed:
reduction on complex numbers, added volatile copy and assignment (#2453)

jjsjann123 committed:
disable index_select/gather tests (#2530)

disables index_select / gather python tests since upstream backs out autodiff support on these ops.
pytorch#95565

We'll re-enable them when we remerge the autodiff support with opt-in via environment variable

naoyam committed:
Fix #2531 (#2532)

* Fix #2531

Changed ReplayTransformations to take bool parameters explicitly with
set methods. This is to avoid accidentally passing those bool arguments
in a wrong order. More verbose, but would be safer

zasdfgbnm committed:
New lowering pass: loop rotation (#2500)

zasdfgbnm committed:
Add optional bool `preserve_error` for expr simplifier (#2534)

zasdfgbnm committed:
Improve matmul instruction scheduling with loop rotation (#2488)

zasdfgbnm committed:
Fix stop predicate (#2537)

jjsjann123 committed:
bump ubuntu version (#2546)

zasdfgbnm committed:
cp.async access global tensor via pointer (#2282)

zasdfgbnm committed:
Rewrite `reducePredicateRegisterUsage` (#2533)

rdspring1 committed:
Fix downcast_ptr != nullptr in RandomOp (#2547)

zasdfgbnm committed:
Update fusion_record.h (#2536)

zasdfgbnm committed:
Make contiguity ignore broadcasts (#2517)

csarofeen committed:
Multidevice setup merged with devel (#2318)

Implements fundamental logic for multi-device support

Co-authored-by: shmsong <shisong@umich.edu>
Co-authored-by: Sergey Lebedev <sergeyle@nvidia.com>
Co-authored-by: snordmann <snordmann@nvidia.com>
Co-authored-by: Xiang Gao <qasdfgtyuiop@gmail.com>
Co-authored-by: Naoya Maruyama <nmaruyama@nvidia.com>

liqiangxl committed:
allow prediction on block dim (#2544)

jjsjann123 committed:
Output stride order (#2548)

Added new python API fd.ops.add_output(tensor, stride_order), where stride_order means that output axis i is the stride_order[i]th fastest dimension.

e.g. if we want to specify output to be in channel-last format, we should specify fd.ops.add_output(tensor_view, [0, 3, 1, 2]), where a given output with shape [N, C, H, W] will have stride [H*W*C, 1, W*C, C]

Implementation details:
It's currently done in a naive way. Since nvfuser doesn't support user specified stride order yet, we fake it by:

    adding a permute op on outputs inside the generated kernel, to ensure that the output is stored in the correct memory layout;
    after the kernel has executed, we permute that corresponding output to undo the permutation inside the kernel, this gives us the semantically correct output in the desired memory layout.

jjsjann123 committed:
skipping ampere tests on pre-ampere devices (#2553)

ampere tests running on pre-ampere devices triggers a CI failure.

zasdfgbnm committed:
Add debugging utility RAII guard for printting scopes (#2555)

mmigdal-nv committed:
Fixing and improving indexing type handling (#2522)

Fixing and improving indexing type handling

jacobhinkle committed:
Avoid adding "f" suffix to std::{real,imag} (#2565)

Fixes #2564

Co-authored-by: Jacob Hinkle <jhinkle@nvidia.com>

naoyam committed:
Revert "Fixing and improving indexing type handling (#2522)" (#2568)

This reverts commit 3b85308a8e303d0df43c2d3cac1edba87dde2e49.

jjsjann123 committed:
Bcast python api patch (#2561)

Fixes python handling of expanded broadcast dimension.
e.g. torch.randint(0, 1, (5, 5), device="cuda").bool().unsqueeze(-1).expand(5, 5, 5)

Changes contiguity representation on python.
computeContiguity currently returns an array with the length of tensor rank, elements in the array can be: True, False or None, where None indicates a given dimension is broadcast.

naoyam committed:
Clean up index type handling (#2570)

* Clean up compile-time and run-time index options

zasdfgbnm committed:
Change contiguity into `std::vector<c10::optional<bool>>` (#2569)

zasdfgbnm committed:
Make expr simplifier tests more readable (#2571)

jjsjann123 committed:
Index select empty tensor scalar tensor (#2513)

Fixes index_select on empty/scalar indices. Issues found in python API.

Our stack should support empty tensor (numel()==0), removed the check on that.
Scalar tensor should be used in place of real Scalar with variable_name[0]. Add a quick patch for that.
---------

Co-authored-by: Naoya Maruyama <nmaruyama@nvidia.com>

naoyam committed:
Fix indexing failure with non-view rfactor (#2562)

* Include non-view rfactor IDs in CA map rfactor ID sets

samnordmann committed:
Fix multidevice tests (#2574)

* fix tests for multicluster fusion

naoyam committed:
IterDomain resize for pad, cat, slice (#2480)

naoyam committed:
persistent_use_of_buffer is accumulated over all the resolution points. (#2576)

Recomputation for each persistent use should be done after the
accumulation is done.

Currently, recomputation and replaceVal can be done redundantly. For
example, on A100, that happens with NvFuserScheduler_BatchNorm_fp32/64/32/256.
